### ISSUE TITLE
Update our TypeScript configuration to match our currently-supported browsers

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#9748]
Our [tsconfig.json](https://github.com/wagtail/wagtail/blob/main/tsconfig.json) is very out of date. This is causing issues as we gradually transition TypeScript. For example, the compiler doesn’t NodeList destructuring (`const elts = [...document.querySelectorAll('a')];`).

---

We need to review the configuration option-by-option, in particular setting `target` to an appropriate value based on our browser support. Ideally, we should also document this in a way that we can do those updates as needed in the future.

Config reference: https://www.typescriptlang.org/tsconfig

Sample tsconfig.json from a Torchbox project as an alternative example:

```jsonc
{
  "compilerOptions": {
    // Webpack will convert ES modules to the required format.
    "module": "esnext",
    "esModuleInterop": true,
    // Setting a very low target for IE11 compatibility.
    "target": "es5",
    // Make it possible to gradually adopt TypeScript, or not use it at all.
    "allowJs": true,
    // Turn on stricter type checking.
    "strict": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": true,
    "sourceMap": true,
    "jsx": "react"
  },
  "include": ["static_src"],
  "exclude": ["node_modules", "static_compiled"]
}
```

---

Once we’ve done our review, we can change the config in Wagtail, making sure the JS keeps working as expected across our supported browsers.

**Comment by thibaudcolas** [comment]
TL;DR; Safari 14 is our likely oldest browser to support, meaning we can use a target of `es2021`.

- Browser support targets for Wagtail: https://docs.wagtail.org/en/stable/contributing/developing.html#browser-and-device-support
- Browser versions table: https://caniuse.com/usage-table
- JS features by version: https://kangax.github.io/compat-table/es2016plus/
